module/os/freebsd/zfs/zfs_vfsops.c

```diff
@@ -89,6 +89,11 @@ int zfs_debug_level;
 SYSCTL_INT(_vfs_zfs, OID_AUTO, debug, CTLFLAG_RWTUN, &zfs_debug_level, 0,
 	"Debug level");
 
+int zfs_jail_mount_snapshot;
+SYSCTL_INT(_vfs_zfs, OID_AUTO, jail_mount_snapshot, CTLFLAG_RW,
```

Comment:
It just occurred to me, what about using `SYSCTL_JAIL_PARAM`?

`cstyle` is flat out broken for multiline macro invocations outside of function scope. IMO it's best to wrap this in `/* BEGIN CSTYLED */ ... /* END CSTYLED */`.

Reply:
SYSCTL_JAIL_PARAM: looking into this, it is an interesting idea, being able to control it on a per-jail basis. If it can be done without requiring modifications on the FreeBSD side (so it will work on 13.1) I really like this idea.

Reply:
So far, it seems like this won't work. Adding the `SYSCTL_JAIL_PARAM()` creates `security.jail.param.zfs.mount_snapshot` but to control it, you'd need to modify `jail` to understand `zfs.mount_snapshot=1`

Basically, without modifying `struct prison`, I am not sure we'd have a good place to store the per-jail setting. I wonder if the OSD stuff might make sense for that.

Reply:
We already use `osd_jail_register()`, `osd_jail_get()`, `osd_jail_set()`, and so on in `spl_zone.c` to extend jails with a list of jailed datasets. Another example that actually adds configurable jail params is FreeBSD's `linux(4)` as seen in `sys/compat/linux/linux_mib.c`. My understanding is that the jail parameters are intended to be dynamic to support use cases such as this. It does seem possible.

Reply:
Under-tested, but it compiles and loads and the params show up. I more or less copied from FreeBSD's sys/compat/linux/linux_mib.c so credit to the people who worked on that over the years.
https://github.com/truenas/zfs/commit/69d2a21a30efbfc95eff701957f3d6985111373c

Reply:
Thanks for doing that, it looks good.
I made one small addition, not allowing a child jail to set mount_snapshot=1 if the parent has it set to 0, since the child should never be able to get more permissions than its parent.